### DISCUSSION TITLE
CUE used as conversion rules for importing data into a ledger

### DISCUSSION THREAD

**Discussion by bttk** [cue-lang/cue#1216] [Ideas]
I am trying to use CUE in what I think is a new way: as a set of rules for ingesting data.

## Background
I keep track of my personal finances using [hledger](https://hledger.org/) - a plain text accounting tool. I've set up a small pipeline which ingests csv files by using [hledger's built-in rules format](https://hledger.org/import-csv.html). Over time I've added enough regex rules for categorizing expenses (approaching 1000 lines), I no longer have confidence the order of rules is correct.

Some rule examples (simplified):
```
;-- checking.rules --
; All transactions debit the checking account
account1 asset:checking

; CSV rows matching any of these patterns credit the dining expenses account
if
\bRESTAURANT\b
\bSUSHI\b
\bPIZZA\b
  account2 expense:dining
```
## Why CUE
CUE seems to have a few properties well suited for this task:
* first-class types can be used to express rule patterns 
* the order of operations shouldn't change the result
* if matching is ambiguous, the value will not be concrete

I'm making progress on getting it to work at all, but I'm not sure if I'll find a way to reduce the amount of boilerplate and use bigint for amounts. Here is an example of an equivalent rule file for now:

```cue
-- checking_rules.cue --
// Names of columns in the csv file
#_rowformat: {
  Date: string
  Desc: string
  Amnt: string
  ...
}

[string]: hledger.#TransactionRule

dining: {
  #row: #_rowformat & {
    Desc: (
      =~"(?i)\brestaurant\b" |
      =~"(?i)\bsushi\b" |
      =~"(?i)\bpizza\b"
    )
  }
  {
    date: #row.Date
    description: #row.Desc
    postings: [{
      account: "asset:checking"
    },{
      account: "expense:dining"
      amount: strconv.ParseFloat(#row.Amnt, 32)
    },]
  }
}
```

**Comment by verdverm** [comment]
Suppose an CSV row would match multiple regex in your existing hldeger rules. Is the reason order is important that hldeger takes the first match?

If so, I don't think that CUE's "order doesn't matter" carries over directly. What happens when there are multiple matches? i.e. the description string has tokens matching multiple transaction rules.

@bttk Could you provide a motivating example where this is the case, if it is indeed a possibility?

**Reply by bttk** [reply]
hledger will execute all of the rules in the order they appear - these indented statements below the patterns are field assignments which overwrite earlier assignments.

One example: I import order history from Amazon. The csv file contains a `category` column, which I find very useful.
```
; the default destination is unkown amazon expenses
account1 expense:unknown:amazon

; items with a category starting with "toy" go into fun expenses
if
%category ^TOY[S_]
  account1 expense:fun

; a toy I bought for lBob
if
%order 123-456890-123456
  account1 expense:gift:bob

; csv file is missing a description for this item, so I append my own note
if
%asin B07238927
  description %description (racing car)
```

When I started I had imagined using CUE's disjunction operator to define a hierarchy of patterns. 
```
-- imagination.cue --
*(
  {
    #row: order: "123-456890-123456"
    txn: account: "expense:gift:bob"
  } | {
    #row: asin: "B07238927"
    txn: description: "append???"
  }
) | (
  #row: category: =~"^toy[s_]"
  txn: account: "expense:fun"
)
```
I'm starting to think it's more feasible to define a hierarchy of structs and to write a tool that will unify the final transaction.

**Reply by verdverm** [reply]
`Disjunction in CUE is a binary expression, written a | b. It is commutative, associative, and idempotent.` (https://cuelang.org/docs/references/spec/#disjunction)

I suspect you will still have to account for regexp ordering. One thing you might consider is generating the hledger rule pipeline from CUE, which could in theory be easier to maintain. Not familiar enough with hledger to comment on how much of the processing can be replaced with CUE

**Comment by myitcv** [comment]
@bttk welcome to the CUE project!

> use bigint for amounts

See [the spec](https://tip.cuelang.org/docs/references/spec/#numeric-values) for more details, but you should be covered on this point.

I'm assuming you're writing this matching engine using CUE's Go API (we don't currently have a reduce-like builtin, which I think prevents you from doing this via `cmd/cue`)? CUE does indeed seem like a very natural to specify rules like this, not least because they could be very easily tested. Certainly happy to help kick ideas around on what such a solution would look like.

**Reply by bttk** [reply]
> @bttk welcome to the CUE project!
👋 
 
> > use bigint for amounts
> 
> See [the spec](https://tip.cuelang.org/docs/references/spec/#numeric-values) for more details, but you should be covered on this point.

So.. if there is no way to extend `number`, there is no accurate representation of `0.1`? For my use this is unlikely to affect me, but still.

> I'm assuming you're writing this matching engine using CUE's Go API (we don't currently have a reduce-like builtin, which I think prevents you from doing this via `cmd/cue`)? CUE does indeed seem like a very natural to specify rules like this, not least because they could be very easily tested. Certainly happy to help kick ideas around on what such a solution would look like.

Yes, I have experimented with the Go API and I'm confident I'll be able to use it, once I decide on the config schema. If you have some ideas, I'm very interested in learning how others would approach it. I see I'll need to change my approach and process records in several passes:

1. normalize raw csv
    * normalize date format
    * normalize amounts, currencies
    * some filtering
2. move values into a transaction struct
3. process individual overrides (by order id or check number)
4. match description by regex to classify the remaining transactions
5. set the defaults for the remaining transactions

You might ask why can't I apply these individual overrides by hand in the plain text ledger. I store all csv inputs and often rerun the pipeline when I change the rules.

**Reply by bttk** [reply]
Update: I'm using the "function" and "switch" patters found on cuetorials.com and this tool to ingest csv data:
https://gist.github.com/bttk/f4596a8143ae35d67f6c5aa71f19aec7

I think I'll be able to do the rest of processing in cue and emit results with a hledger plain-text format.

**Reply by bttk** [reply]
Update 2: I've been incrementally working on this. Cuetorials and youtube streams were a great learning resource! Here are a few cue-isms which may be useful in other places:

.csv files are converted into .cue files matching this schema:
```cue
// This is inspired by the schema of *_tool.cue
table: [string]: #Content

#Content: {
	// Column names used when parsing the file.
	header: [...string]

	rows: [...{
		// Row as a struct:
		// { [columnName]: columnValue }
		[string]: string
	}]

	// Ensure column names are unique.
	_uniq_header: {
		for i, v in header {
			if v != "" {
				"\(v)": i
			}
		}
	}

	// Ensure fields in rows correspond to column names.
	_valid_rows: [
		for row in rows {
			for k, v in row {
				"\(k)": _uniq_header["\(k)"]
			}
		},
	]
}
```

Using a similar schema I define functions for processing rows.
```cue
func: [string]: #Func

#Func: {
	#in: _
	_
}

// Returns a list
#ListMapper: #Func & {
	IN=#in: [...]
	FN=#fn: #Func

	[ for r in IN {
		FN & {#in: r, _}
	}]
}

// Returns a string
#TextMapper: #Func & {
	IN=#in: [...]
	FN=#fn: #Func
	#join:  string | *"\n"

	#_list: #ListMapper & {#in: IN, #fn: FN, [...string]}
	strings.Join(#_list, #join)
}
```

The third ingredient is a utility to execute any `func` on any `table`:
```cue
// apply.cue
package apply

// import ":mapper"

func: [string]: mapper.#Func
table: [string]: rows: [...]

mapperType: "text" | *"list"    @tag(mapper)
tableName:  string  @tag(table)
funcName:   string  @tag(func)

if mapperType == "list" {
	result: mapper.#ListMapper & {
		#in: table[tableName].rows
		#fn: func[funcName]
		_
	}
}

if mapperType == "text" {
	result: mapper.#TextMapper & {
		#in: table[tableName].rows
		#fn: func[funcName]
		_
	}
}
```

Everything is then put together with a single command:
`cue export ./foo_table.cue ./bar_func.cue ./apply.cue -e result --inject table=foo --inject func=bar`

It's important that `foo_table.cue` and `bar_func.cue` don't define a package name.
This is where I'm using a 70-LOC cue func transforming a transaction struct into the plain-text ledger format.

The last thing is a util for concatenating table rows:
```cue
package concat

import (
	"list"
	"strings"
)

DATA=table: [string]: rows: [...]

// A comma-separated list of table names. Default: all tables.
tableNames: string | *null     @tag(tables)
resultName: string | *"concat" @tag(name)

result: {
	#_tableKeys: [
			if tableNames != null {
			strings.Split(tableNames, ",")
		},
		[ for k, v in DATA {k}],
	][0]

	table: "\(resultName)": rows: list.FlattenN([
		for _, k in #_tableKeys {
			DATA[k].rows
		}], 1)
}
```
